lib/spack/spack/variant.py

```diff
@@ -617,6 +623,102 @@ def substitute_abstract_variants(spec):
         spec.variants.substitute(new_variant)
 
 
+class DisjointSetsOfValues(Sequence):
+    """Allows combinations from one of among many mutually exclusive sets.
+
+    Args:
+        *sets (list of tuples): mutually exclusive sets of values
+    """
+    def __init__(self, *sets):
+        self.sets = [set(x) for x in sets]
+        if any(s1 & s2 for s1, s2 in itertools.combinations(self.sets, 2)):
+            raise error.SpecError('sets in input must be disjoint')
+
+        self.feature_values = tuple(itertools.chain.from_iterable(self.sets))
+        self.default = None
+        self.multi = True
+        self.error_fmt = "variant '{0}' in package '{1}' accepts " \
+                         "combinations from a single disjoint set of " \
+                         "values [{2}]"
+
+    def with_default(self, default):
+        """Sets the default value and returns self."""
+        self.default = default
+        return self
+
+    def with_error(self, error_fmt):
+        """Sets the error message format and returns self."""
+        self.error_fmt = error_fmt
+        return self
+
+    def with_non_feature_values(self, *values):
+        """Marks a few values as not being tied to a feature."""
+        self.feature_values = tuple(
+            x for x in self.feature_values if x not in values
+        )
+        return self
+
+    def __contains__(self, item):
+        return item in itertools.chain.from_iterable(self.sets)
+
+    def __getitem__(self, idx):
```

Comment:
> I am certainly biased, but I think I've made a stronger case when we've discussed it before than just stating my opinion, and I also did that here (in a shorter format). 

So did I, I think (the reference to our previous discussion is https://github.com/spack/spack/pull/4307#discussion_r131808515). Briefly my points are:
1. `collections.abc` is  a standard library module, not some exotic third-party package.
2. Deriving from `Sequence` to code an object that mimics a `tuple` is an accepted idiom in the Python community (as it is deriving from `MutableMapping` to mimic a `dict`).
3. It's not always true that less functionality means less complexity. Probably there are more Python users that would be comfortable with the concept of `tuple` than with that of an "iterable container" (i.e. an object that implements `__iter__` and `__contains__`).

Sorry if I insist on this, but implementing just `__iter__` and `__contains__` feels like introducing a bad practice to me. Also, the changes to implement just those two methods will give more or less the same line count as it is now - so no added benefit on that front.

That said, given that my primary interest here is to fix a bug + extend the functionality we provide to packagers, if for you having `Sequence` is enough to prevent this PR from being merged I'll change the implementation on your request. In that case can you please suggest how to reword:
https://github.com/spack/spack/blob/62d821d8431f5430484a95927cd5c0a882f808ff/lib/spack/spack/directives.py#L462

as we would break the promise to accept a tuple, and somebody extending the `variant` directive might get hit by this discrepancy in the future.

Reply:
> `collections.abc` is a standard library module, not some exotic third-party package.

None of my complaints have been about `collections.abc` being third-party or obscure. Regardless of how core it is to the language, I still think it increases the complexity.

> Deriving from `Sequence` to code an object that mimics a `tuple` is an accepted idiom in the Python community

I'm not sure how to assess the value of this: how much of the Python community accepts it? Why do they accept it? Presumably there are downsides, for one this isn't interchangeable with a tuple because it doesn't implement `__hash__`. For future reference I prefer either a link to a document I can read, or an explanation of the reasons why the Python community would prefer this.

> It's not always true that less functionality means less complexity. Probably there are more Python users that would be comfortable with the concept of tuple than with that of an "iterable container" (i.e. an object that implements __iter__ and __contains__)

This highlights a disconnect between the internals of the function (and what it is capable of handling) and the documentation that should be provided to users. I agree that it is likely best to tell users "use a tuple" because that works and is what they are familiar with. From the perspective of those editing `Variant`, I expect `__iter__` to be understandable. More on this with the reword suggestion below.

Also, I'm not sure why you need to implement `__contains__` separately: that often comes "for free" with `__iter__`.

> can you please suggest how to reword:

'values' can be any object with an idempotent `__iter__` method, or a callable. As far as the user-facing aspect of the docstring goes, this is likely to be too much detail. I'd keep the docstring the same (since tuples *are* acceptable), and add a `#`-style comment below.

I'd summarize by saying that I'm still not convinced, *BUT* this class is small and self-contained enough (unlike `RepoPath` in #4307) that I'm considering letting it go. I think that some of your motivation comes from merging usage concerns and maintenance concerns, and I think at least some of this is better addressed with documentation. Minimally, I request that you add a comment near the class declaration which mentions that you are extending `Sequence` to give this object the appearance of a tuple.

When core devs put effort into fitting the smallest reasonable interface, it saves me headaches worrying about whether this will set a precedent for other Spack coders to start extending container classes (again, I still think it's not a good choice for `RepoPath`).